File: handarm/ik.py
"""Damped least squares (Levenberg-Marquardt) inverse kinematics.

Solves for the 6 arm joint angles that place the tool center point at a
target pose. Three practical concerns drive the design:

* Singularities. Near a singular configuration the Jacobian loses rank and
  a plain pseudo-inverse commands huge joint velocities. We monitor the
  manipulability measure w = sqrt(det(J J^T)) and ramp extra damping in as
  w collapses, trading tracking accuracy for bounded, stable steps.

* Joint limits. Every iterate is clamped to the URDF/MJCF joint ranges, so
  the solver only ever explores feasible configurations.

* Continuity. The solver warm-starts from the current configuration and
  clamps both the task-space error and the per-iteration joint step, so
  consecutive frames produce nearby solutions instead of branch flips.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np

from .config import IKConfig
from .transforms import orientation_error


@dataclass
class IKResult:
    q: np.ndarray
    pos_err: float   # meters
    ori_err: float   # radians
    iters: int
    converged: bool
    manipulability: float


class DLSSolver:
    def __init__(self, kin, cfg: IKConfig | None = None):
        """kin: object with fk(q), jacobian(q), clamp(q), lower/upper."""
        self.kin = kin
        self.cfg = cfg or IKConfig()

    def solve(self, q_init: np.ndarray, target_pos: np.ndarray,
              target_quat: np.ndarray, restarts: int = 0) -> IKResult:
        """Solve IK warm-started from q_init.

        restarts > 0 adds cold-start attempts (base yaw aimed at the
        target, then random seeds) with a larger iteration budget; used
        for one-shot solves, not the per-frame tracking path.
        """
        best = self._descend(q_init, target_pos, target_quat, self.cfg.max_iters)
        if best.converged or restarts <= 0:
            return best

        seeds = [self._aimed_seed(q_init, target_pos)]
        rng = np.random.default_rng(0)
        while len(seeds) < restarts:
            seeds.append(self.kin.lower + rng.random(self.kin.ndof)
                         * (self.kin.upper - self.kin.lower))
        for seed in seeds:
            res = self._descend(seed, target_pos, target_quat, self.cfg.max_iters * 4)
            if res.converged:
                return res
            if res.pos_err + res.ori_err < best.pos_err + best.ori_err:
                best = res
        return best

    def _aimed_seed(self, q_init: np.ndarray, target_pos: np.ndarray) -> np.ndarray:
        q = np.asarray(q_init, dtype=float).copy()
        q[0] = np.arctan2(target_pos[1], target_pos[0])
        return q

    def _descend(self, q_init: np.ndarray, target_pos: np.ndarray,
                 target_quat: np.ndarray, max_iters: int) -> IKResult:
        cfg = self.cfg
        q = self.kin.clamp(np.asarray(q_init, dtype=float).copy())
        pos_err = ori_err = np.inf
        manip = 0.0
        it = 0

        for it in range(1, max_iters + 1):
            pos, quat, J = self.kin.fk_and_jacobian(q)
            # Manipulability (Nakamura & Hanafusa), measured on the raw
            # Jacobian before task weighting so the damping threshold has
            # a fixed geometric meaning.
            manip = float(np.sqrt(max(np.linalg.det(J @ J.T), 0.0)))

            e_pos = target_pos - pos
            e_ori = orientation_error(target_quat, quat)
            pos_err = float(np.linalg.norm(e_pos))
            ori_err = float(np.linalg.norm(e_ori))
            if pos_err < cfg.pos_tol and ori_err < cfg.ori_tol:
                return IKResult(q, pos_err, ori_err, it, True, manip)

            # Clamp the translational error so distant targets pull the
            # arm smoothly instead of demanding one giant step.
            if pos_err > cfg.max_pos_err:
                e_pos = e_pos * (cfg.max_pos_err / pos_err)
            e = np.concatenate([e_pos, cfg.ori_weight * e_ori])

            J = np.vstack([J[:3], cfg.ori_weight * J[3:]])
            JJT = J @ J.T
            lam = cfg.base_damping
            if manip < cfg.manip_threshold:
                lam += cfg.singular_damping * (1.0 - manip / cfg.manip_threshold)

            dq = J.T @ np.linalg.solve(JJT + (lam ** 2) * np.eye(6), e)
            dq = np.clip(dq, -cfg.max_step, cfg.max_step)
            q = self.kin.clamp(q + dq)

        pos, quat = self.kin.fk(q)
        pos_err = float(np.linalg.norm(target_pos - pos))
        ori_err = float(np.linalg.norm(orientation_error(target_quat, quat)))
        converged = pos_err < cfg.pos_tol and ori_err < cfg.ori_tol
        return IKResult(q, pos_err, ori_err, it, converged, manip)

File: web/src/ik.ts
/**
 * Damped least squares (Levenberg-Marquardt) inverse kinematics —
 * TypeScript port of handarm/ik.py.
 *
 * dq = J^T (J J^T + lambda^2 I)^-1 e, with manipulability-adaptive
 * damping near singularities, joint-limit clamping on every iterate, and
 * clamped error/step magnitudes for frame-to-frame continuity.
 */

import { IK } from "./config";
import { clampJoints, fk, jacobian } from "./kinematics";
import { orientationError, type Quat, type Vec3, vecNorm } from "./transforms";

export interface IKResult {
  q: number[];
  posErr: number; // meters
  oriErr: number; // radians
  iters: number;
  converged: boolean;
  manipulability: number;
}

/**
 * Solve A x = b for a 6x6 system via Gaussian elimination with partial
 * pivoting. Also returns det(A) from the pivot product.
 */
export function solveLinear6(
  A: number[][],
  b: number[],
): { x: number[]; det: number } {
  const n = 6;
  const M = A.map((row, i) => [...row, b[i]]);
  let det = 1;
  for (let col = 0; col < n; col++) {
    let pivot = col;
    for (let r = col + 1; r < n; r++) {
      if (Math.abs(M[r][col]) > Math.abs(M[pivot][col])) pivot = r;
    }
    if (pivot !== col) {
      [M[col], M[pivot]] = [M[pivot], M[col]];
      det = -det;
    }
    const p = M[col][col];
    det *= p;
    if (Math.abs(p) < 1e-14) return { x: new Array(n).fill(0), det: 0 };
    for (let r = col + 1; r < n; r++) {
      const f = M[r][col] / p;
      for (let c = col; c <= n; c++) M[r][c] -= f * M[col][c];
    }
  }
  const x = new Array(n).fill(0);
  for (let r = n - 1; r >= 0; r--) {
    let s = M[r][n];
    for (let c = r + 1; c < n; c++) s -= M[r][c] * x[c];
    x[r] = s / M[r][r];
  }
  return { x, det };
}

function matMulT(J: number[][]): number[][] {
  // J (6x6) times its transpose.
  const out: number[][] = Array.from({ length: 6 }, () => new Array(6).fill(0));
  for (let i = 0; i < 6; i++)
    for (let j = 0; j < 6; j++) {
      let s = 0;
      for (let k = 0; k < 6; k++) s += J[i][k] * J[j][k];
      out[i][j] = s;
    }
  return out;
}

function detOnly(A: number[][]): number {
  return solveLinear6(A, new Array(6).fill(0)).det;
}

function descend(
  qInit: number[],
  targetPos: Vec3,
  targetQuat: Quat,
  maxIters: number,
): IKResult {
  let q = clampJoints([...qInit]);
  let posErr = Infinity;
  let oriErr = Infinity;
  let manip = 0;
  let it = 0;

  for (it = 1; it <= maxIters; it++) {
    const f = fk(q);
    const J = jacobian(q, f);
    // Manipulability (Nakamura & Hanafusa) on the raw Jacobian so the
    // damping threshold has a fixed geometric meaning.
    manip = Math.sqrt(Math.max(detOnly(matMulT(J)), 0));

    let ePos: Vec3 = [
      targetPos[0] - f.pos[0],
      targetPos[1] - f.pos[1],
      targetPos[2] - f.pos[2],
    ];
    const eOri = orientationError(targetQuat, f.quat);
    posErr = vecNorm(ePos);
    oriErr = vecNorm(eOri);
    if (posErr < IK.posTol && oriErr < IK.oriTol) {
      return { q, posErr, oriErr, iters: it, converged: true, manipulability: manip };
    }

    // Clamp the translational error so distant targets pull smoothly.
    if (posErr > IK.maxPosErr) {
      const s = IK.maxPosErr / posErr;
      ePos = [ePos[0] * s, ePos[1] * s, ePos[2] * s];
    }
    const e = [...ePos, ...eOri.map((v) => v * IK.oriWeight)];

    // Weight the orientation rows.
    const Jw = J.map((row, r) =>
      r < 3 ? [...row] : row.map((v) => v * IK.oriWeight),
    );
    const JJT = matMulT(Jw);
    let lam = IK.baseDamping;
    if (manip < IK.manipThreshold) {
      lam += IK.singularDamping * (1 - manip / IK.manipThreshold);
    }
    for (let i = 0; i < 6; i++) JJT[i][i] += lam * lam;

    const { x: y } = solveLinear6(JJT, e);
    // dq = Jw^T y, clamped per joint.
    const dq = new Array(6).fill(0);
    for (let i = 0; i < 6; i++) {
      let s = 0;
      for (let r = 0; r < 6; r++) s += Jw[r][i] * y[r];
      dq[i] = Math.min(Math.max(s, -IK.maxStep), IK.maxStep);
    }
    q = clampJoints(q.map((v, i) => v + dq[i]));
  }

  const f = fk(q);
  posErr = vecNorm([
    targetPos[0] - f.pos[0],
    targetPos[1] - f.pos[1],
    targetPos[2] - f.pos[2],
  ]);
  oriErr = vecNorm(orientationError(targetQuat, f.quat));
  const converged = posErr < IK.posTol && oriErr < IK.oriTol;
  return { q, posErr, oriErr, iters: it - 1, converged, manipulability: manip };
}

/** Warm-started per-frame solve (the tracking path). */
export function solveIK(
  qInit: number[],
  targetPos: Vec3,
  targetQuat: Quat,
): IKResult {
  return descend(qInit, targetPos, targetQuat, IK.maxIters);
}
